### ISSUE TITLE
IDE should drop queued parsing requests for editing

### ISSUE THREAD

**Issue by OmarTawfik** [dotnet/roslyn#26767]
IDE currently queues a parsing request for every key hit. For large files, when parsing can take more than a few milliseconds, these requests queue fast enough, causing unresponsiveness in the editor.
The IDE should be able to clear the queue once additional input is available from the user (making the queue obsolete).

**Comment by CyrusNajmabadi** [comment]
@omartawfik, can you provide more information on this?  How is this causing unresponsiveness in the editor? These scenarios have been heavily profiled and this work should be happening no the BG.  This should not be impacting responsiveness.

**Comment by OmarTawfik** [comment]
I'll let @heejaechang provide the additional details. He had a couple of ideas about doing that in cases where there is a huge file (10ks of lines) with parse errors, so a lot of nodes get re-parsed.

**Comment by CyrusNajmabadi** [comment]
> so a lot of nodes get re-parsed.

I'd like to understand this better :)   How is that leading to a problem.  Note that errors only cause hte spine up the node-with-error to be considered unusable.  All the non-errors nodes around it should be fine.

I'm also a bit unsure how we would actually accomplish this cancellation.  Say you have 10 features that woke up and are asking for that tree.  How do you 'cancel' the parse?  What will be returned to all those features?  

Another major concern to me is that if we cancel, and then reparse and cancel and reparse over and over again, we'll just be repeating the same work over and over again.  One of hte benefits of hte existing system is that by allowing hte previous to make it to completion, we can then use that new tree as the basis of further trees/changes.   

@heejae, before we proceed on these points, i'd really like to understand the core scenario and hte data/analysis around it.  It may be that it warrants this sort of design change.  But i would lke to understand hte data better as it may indicate a problem elsewhere entirely.

**Comment by sharwell** [comment]
@OmarTawfik Ideally, when creating a performance item like this, it starts with a testing scenario or other case study capable of demonstrating the problem and manually verifying lack of regressions. See #23583 for an example of one of these scenarios.

**Comment by heejaechang** [comment]
it is open to any idea. 

this is fact that we chain incremental parsing request blindly one by one in compilation tracker. and it is more efficient to do 1 incremental parsing of multiple changes if there are close by rather than doing multiple incremental parsing of 1 char changes each.

this is also a fact that incremental parsing sometimes take longer than we hope to.

whether those 2 requires improvement or not is open question and if we do, how is also another often question.

**Comment by heejaechang** [comment]
tagging @jinujoseph

**Comment by jinujoseph** [comment]
Talking to DTS Customer with analysis results and getting a sense of importance

**Comment by CyrusNajmabadi** [comment]
Have we investigated incremental parsing to see if we're hitting some issue htere?  I did extensive perf testing of this over in TS land when we did the parser there.  That was a much harder parse to make fast, given that we were ultimately running JS code in older engines. But incremental parsing was not a problem, even on large files (10k+ lines).  

Are you sure you're not hitting some sort of edge case in the parser itself and are trying to attack the problem from a less optimal direction?

--

Note: no data or test cases have been presented yet that actually demonstrate the problem.  So it's really hard to tell what's going on, or what might be the best thing to do about it without that.

**Comment by CyrusNajmabadi** [comment]
> this is fact that we chain incremental parsing request blindly one by one in compilation tracker. and it is more efficient to do 1 incremental parsing of multiple changes if there are close by rather than doing multiple incremental parsing of 1 char changes each.

See my messages in https://github.com/dotnet/roslyn/issues/26766 for my concerns here.  

--

Could more data be presented about what sort of issue the customer is actually running into?  @OmarTawfik said there is an issue that is "causing unresponsiveness in the editor."   That really needs to be understood.  How is this causing unresponsiveness in the editor?  Without a clear understanding of hte actual problem (from end to end) it's very hard to make any statements about an appropriate course of action.

**Comment by heejaechang** [comment]
in this particular bug, we see there are about 50 incremental parse tree chained for the document user is trying to use. users is blocked to get syntax tree of the document since we are waiting all those 50 chained incremental parsing to happen.

if our design didn't depends on previous tree to get current tree, user could just get current tree without waiting all those 50 trees to be parsed.

**Comment by CyrusNajmabadi** [comment]
interesting.  So they've been albe to make 50 edits without a problem, but then they're blocked on one op.  Do we know what hte one op is that is causing a problem for them with the previous 50 ops were not a problem?

**Comment by heejaechang** [comment]
that is not how we work. we are lazy, we keep chain work until someone tries to use it. user tries to get tree of a document, and it has 50 pending incremental parsing work pending for the document. and to get current tree, we need to parse all those.

**Comment by CyrusNajmabadi** [comment]
>how we do that is up to discussion. for example, we could make current blind chain to something like GetBaseTreeAndTextChanges and then based on returned tree and accumulated text changes to get current tree.

That idea seems very promising to me.  The 'blind chaining' approach only seems to make sense if things are fast.  Though things *should* be fast, so we do need to understand why it isn't.   However, a dual pronged attack of addressing fundamental perf issues, and then making the design not suffer if hte underlying perf is bad, would make sense.

**Comment by CyrusNajmabadi** [comment]
> that is not how we work. we are lazy, we keep chain work until someone tries to use it. user tries to get tree of a document, and it has 50 pending incremental parsing work pending for the document. and to get current tree, we need to parse all those. that means, no one requested parse tree for those 50 edits.

Something is broken then.  That was not our design for the VS workspace.  That's precisely why BackgroundParser works.  The goal if it is to *precisely* ensure that our trees are up to date and to make it so we do not have long chains.

**Comment by heejaechang** [comment]
background parsing is not caught up with lastest tree as well for the document. due to the design we choose, we have background parser to make it up to date as soon as possible (so we don't have long chain hopefully). but in this particular case, it also has pending works, but didn't finish processing them yet.

in this kind of situation, BG parsing seems not helpful as well. since it will either get cancelled before finishing parse tree (shortening the chain length), or get blocked same as the users.

**Comment by CyrusNajmabadi** [comment]
> BG parsing seems not helpful as well. 

That seems unfortunately broken since it's failing at the precise job it was meant for :D  Perhaps it should also be fixed up as part of any work here.

**Comment by CyrusNajmabadi** [comment]
> BG parsing seems not helpful as well. since it will either get cancelled before finishing parse tree

Would this matter though?  As omar's other issue stated, incremental parsing is not cancellable.  So who cares if the BG parsing cancels or not?  

Is hte bug simply htat BGParser cancels, and then the actual parse result (which seems like it will complete) doesn't actualy get recorded?  Thus the next BGParse can't use that?

**Comment by heejaechang** [comment]
our chaining has nothing to do with compiler side. the cancellation will make the chain not to be shorted.

**Comment by CyrusNajmabadi** [comment]
Note: i've tried to understand all the ValueSource code here and how all these pieces of data get stored once computed... and i can't make heads or tails of it.  It wouldn't surprise me at all to learn that we were doing all the necessary work, and then just losing it, forcing the computation 'down hte line' for things we already figured out and could have held onto if the system wasn't so complex here :-/

**Comment by CyrusNajmabadi** [comment]
> I dont think waiting for non cancellable work when cancellation is raised is right answer as well.

That's not waht i was saying.  I'm not saying you should have to wait for the work.  I'm syaing that the work is still happening, and when it *finishes*, it is not beneficial to throw that work away.  In the 'sequence of edits' case, that just means you're going to repeat the same work over and over again (see my E1-E10 description.

**Comment by CyrusNajmabadi** [comment]
> hmmm.. but that's how task works. 

Parsing (and incremental parsing) isn't a Task.  It's a straight synchronous call that concludes with a resultant value.  The question is: what do we do with that value?  If the *top level* caller cancelled, do we lose that value?  If so, that just hurts us because we pay the price N times.  We have to pay the price to parse anyways, and then we pay the price over and over again because BGParser is 'cancelling' but then having to have it happen again when it kicks off hte next parse.  That can quickly explode into a problem because BGParser isn't actually keeping up with the edits.  Instead, it's just cancelling its existing work, and then starting over *from the beginning*. 

This violates the goals of incremental parsing and BGParsing, which should generally only be "i have a few edits to keep up with as typing is happening".

**Comment by heejaechang** [comment]
> For things like ValueSource if your caller cancels, you can return right away to them, but that doesn't mean you can't still take the resultant work that completes and store that for the next caller.

sure, we can hack around as much as we want to. but I think that will be new concept as well. also, not sure why incremental  parsing is not cancellable at the first place. there is no gurantee that changes for incremental parsing is small (since one can replace whole file or most of it)

**Comment by heejaechang** [comment]
this seems not good that there is no cancellation on incremental parsing. when full parse has. especially if we are saying WithChange can cause full parsing some times.

/// <summary>
        /// Create a new syntax tree based off this tree using a new source text.
        /// 
        /// If the new source text is a minor change from the current source text an incremental
        /// parse will occur reusing most of the current syntax tree internal data.  Otherwise, a
        /// full parse will occur using the new source text.
        /// </summary>
        public abstract SyntaxTree WithChangedText(SourceText newText);

....

/// <summary>
        /// Produces a syntax tree by parsing the source text.
        /// </summary>
        public static SyntaxTree ParseText(
            string text,
            CSharpParseOptions options = null,
            string path = "",
            Encoding encoding = null,
            CancellationToken cancellationToken = default(CancellationToken))

**Comment by CyrusNajmabadi** [comment]
> sure, we can hack around as much as we want to

This is not a hack.  It's the intent of the system.  It's literally "incremental".  Incremental works when you make small increments and you keep track of those, so that all future work is still small.  Incremental doesn't work if you both do the work *then* throw it away, then do it again.

This was the intended design.  I think we've def found a problem here, but i think we should address that problem.

> also, not sure why incremental parsing is not cancellable at the first place

I have gone into depth on that in the other report by @OmarTawfik .  These were explicit design decisions and the system was intended to work a very specific way with that.

**Comment by CyrusNajmabadi** [comment]
> this seems not good that there is no cancellation on incremental parsing. when full parse has. especially if we are saying WithChange can cause full parsing some times.

this is incorrect.  Incremental parsing should really *not* be cancellable (and it would likely be a mistake to try to make it so).  Because the point of this system is to allow you to precisely chain all those bits of small work easily so that they're ready and available for all the parts of the system that need it.  

Please seem my statements over in https://github.com/dotnet/roslyn/issues/26766

Again, these were explicit design decisions.  but the IDE code seems to have screwed up here somewhat by making what should be a clean, linear algorithm into an unintentional N^2 algorithm.  We should not put a fix in an inappropriate location because it masks a problem with how we wrote our higher layer on top of the lower layer which is working as it was intentionally designed.

**Comment by heejaechang** [comment]
but API doesn't let you dictate whether WithChangedText do incremental parsing or full parsing. what happen it decides to do full parsing and not cancellable? 

whether incremental parsing actually do cancellation or not is implementation detail. or is it caller's responsibility to guess whether WithChangedText will do incremental parsing or full parsing and call ParseText + cancellation token instead?

**Comment by CyrusNajmabadi** [comment]
> whether incremental parsing actually do cancellation or not is implementation detail. or is it caller's responsibility to guess whether WithChangedText will do incremental parsing or full parsing and call ParseText + cancellation token instead?

No.  The callers responsibility is to not ask the compiler to do the same work over and over again.  Imagine a bug where we actually *never* stored a compiler result and we always parsed using all edits from the poitn the user started VS.  That would clearly be understood to be a bug and unacceptable.  We wouldn't say that incremental parsing was busted, or that it needed to be cancellable.  We would say: "it makes no sense to perform all those edits from the start of time, including up to right before hte last edit, and then throw it all ways".

It is explicitly not using the compiler API as we originally designed it to be used.  When we encounter that sort of issue, the solution is not to throw out everything and say we need a new design and contract.  The solutoin is to fix up our side to use things as per the designs that were laid out and agreed on.

Note: this may be why TS does not have a similar issue.  When we're computing intermediary states, we always store the result, and we always use that latest stored result (no matter what) in order to then compute the next delta.

A core design point of our algorithms is that intermediate values are not trash.  They're a critical part of the design philosophy and should be used effectively to get the best perf out of the system.

**Comment by CyrusNajmabadi** [comment]
Closing out.  It's not clear how to make this scenario work at all.  As mentioned in the linked issue, we need these intermediary parsed syntax trees as we have lots of worked kicked off in-flight that wil produce incremental results on them.  it's definitley *not* teh intent of the system to only have a 'latest' tree that all work restarts on.
